doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+            callback into the cpu cache (generates cache misses for all the other callback 
+            elements in the vector which is being iterated). In the callback a method in an 
+            object is called (the object has to be loaded again into the cpu cache, cache miss for 
+            every member which is not used by the method).
+
+    **Potential optimization:**
+      - Create an additional bool member `m_wasTriggered` which is true when at least 
+        one bool in `m_triggerdBy` is true. This would spare us from iterating over the array 
+        when it was not triggered at all.
+
+    **Adjustments:**
+      - Create the class with a template bool array size argument to be more flexible and memory efficient.
+
+    **Note:**
+      - Sadly, we cannot pursue this approach in the WaitSet since it is event 
+        and state based at the same time and this approach supports only event 
+        driven triggering.
+
+  - add class `ConditionVariableWithOriginTracingSignaler` (I know, better name)
+  
+    **Reason:** it behaves quiete differently then the `ConditionVariable`. The `ConditionVariable` tells you:
+    someone has signalled you and you have to find out who it was. The `ConditionVariableWithOriginTracing`
+    would tell you: A, B and C have signalled you.
+
+    ```cpp
+    class ConditionVariableWithOriginTracingSignaler {
+      public:
+        // originId = corresponds to the id in the ConditionVariable atomic_bool array to 
+        // identify the trigger origin easily
+        void notify(const uint64_t originId); 
+
+        // alternative: origin is set in the ctor and we just call notify
+        ConditionVariableWithOriginTracingSignaler(const uint64_t originId);
+        void notify();
+    };
+    ```
+
+  - add class `ConditionVariableWithOriginTracingWaiter`
+  ```cpp
+  class ConditionVariableWithOriginTracingWaiter {
+    public:
+      // returns a list of ids which have notified the conditionVariable
+      cxx::vector<uint64_t> wait(); 
+      // sets the atomic bool to false in m_triggeredBy[id]
+      void resetWaitState(const uint64_t id) noexcept; 
+  };
+  ```
+
+#### ReactAl 
+```cpp
+enum class ReactAlErrors {
+  CALLBACK_CAPACITY_EXCEEDED
+};
+
+template<uint64_t CallbackCapacity>
+class ReactAl {
+  public:
+    template<typename EventOrigin, typename EventType>
+    cxx::expected<ReactAlErrors> attachEvent(
+      EventOrigin & origin, const EventType & eventType,
+      const cxx::function_ref<void(EventOrigin&)> & callback
+    ) noexcept;
+
+    template<typename EventOrigin, typename EventType>
+    void detachEvent( EventOrigin & origin, const EventType & eventType );
+};
+
+```
+ 
+  - Contains a thread which will wake up and iterate through ConditionVariable
+    bool array. If an entry is true, set it to false and then call the 
+    corresponding callback.
+
+#### Class which is attachable to ReactAl 
```

Comment:
The restrictions on the user defined type using this pattern need to be clarified. What is the purpose of those entities and what must be done/provided by the client class internally and what can the user of the class do (this is shown somewhat in the usage above).

Still, from this alone I would have a hard time to write such a user defined class.

Reply:
I will add this in detail when the implementation is done. At the moment this document should provide you with a first glimpse on where the journey is going.